staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc_test.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flowcontrol
+
+import (
+	"math"
+	"math/rand"
+	"sort"
+	"testing"
+)
+
+// TestConcAlloc tests computeConcurrencyAllocation with a bunch of randomly generated cases.
+func TestConcAlloc(t *testing.T) {
+	rands := rand.New(rand.NewSource(1234567890))
+	for i := 0; i < 10000; i++ {
+		probLen := ([]int{0, 1, 2, 3, 4, 6, 9})[rands.Intn(7)]
+		classes := make([]allocProblemItem, probLen)
+		var lowSum, highSum float32
+		var requiredSum int
+		var requiredSumF float32
+		style := "empty"
+		if probLen > 0 {
+			switch rands.Intn(20) {
+			case 0:
+				style = "bound from below"
+				requiredSum = rands.Intn(probLen * 3)
+				requiredSumF = float32(requiredSum)
+				partition32(rands, probLen, requiredSumF, func(j int, x float32) {
+					classes[j].lowerBound = x
+					classes[j].target = x + 2*rands.Float32()
+					classes[j].upperBound = x + 3*rands.Float32()
+					lowSum += classes[j].lowerBound
+					highSum += classes[j].upperBound
+				})
+			case 1:
+				style = "bound from above"
+				requiredSum = rands.Intn(probLen*3) + 1
+				requiredSumF = float32(requiredSum)
+				partition32(rands, probLen, requiredSumF, func(j int, x float32) {
+					classes[j].upperBound = x
+					classes[j].lowerBound = x * f32max(0, 1.25*rands.Float32()-0.25)
+					classes[j].target = classes[j].lowerBound + rands.Float32()
+					lowSum += classes[j].lowerBound
+					highSum += classes[j].upperBound
+				})
+			default:
+				style = "general"
+				for j := 0; j < probLen; j++ {
+					x := f32max(0, rands.Float32()*5-1)
+					classes[j].lowerBound = x
+					classes[j].target = x + 2*rands.Float32()
+					classes[j].upperBound = x + 3*rands.Float32()
```

Comment:
This could produce a target greater than the upperBound?

Reply:
Yes, that is an allowed case.